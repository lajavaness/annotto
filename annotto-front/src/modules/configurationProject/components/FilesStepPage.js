import { Alert, Button, Form, Modal, Space, Upload, message } from 'antd'
import { UploadOutlined } from '@ant-design/icons'
import { debounce, isEmpty, isEqual, some } from 'lodash'
import { useDispatch, useSelector } from 'react-redux'
import { useTranslation } from 'react-i18next'
import React, { useCallback, useEffect } from 'react'

import {
	updateConfigProject,
	uploadAnnotations,
	uploadItems,
	uploadPredictions,
} from 'modules/configurationProject/actions/configurationProjectActions'

import {
	selectConfigAnnotations,
	selectConfigItems,
	selectConfigPredictions,
	selectConfigProject,
} from 'modules/configurationProject/selectors/configurationProjectSelectors'
import { selectIsSimilarityAllowed } from 'modules/root/selectors/rootSelectors'

import {
	ANNOTATIONS,
	API,
	CONFIG,
	DEADLINE,
	FILE,
	ITEMS,
	PREDICTIONS,
	PROJECT,
} from 'shared/enums/configurationsProjectPropertiesTypes'

import * as Styled from 'modules/configurationProject/components/__styles__/FilesStepPage.styles'

const FilesStepPage = ({ isEdit }) => {
	const projectConfig = useSelector(selectConfigProject)
	const itemsConfig = useSelector(selectConfigItems)
	const predictionsConfig = useSelector(selectConfigPredictions)
	const annotationsConfig = useSelector(selectConfigAnnotations)
	const isSimilarityAllowed = useSelector(selectIsSimilarityAllowed)

	const dispatch = useDispatch()

	const { t } = useTranslation('configurationProject')

	const [form] = Form.useForm()

	useEffect(() => {
		if (projectConfig) {
			form.setFieldsValue(projectConfig)
		}
	}, [form, projectConfig])

	useEffect(() => {
		if (projectConfig || predictionsConfig || itemsConfig?.file) {
			const formData = {
				api: projectConfig?.[API],
				[CONFIG]: projectConfig?.file,
				[PREDICTIONS]: predictionsConfig && [predictionsConfig],
				[ANNOTATIONS]: annotationsConfig && [annotationsConfig],
				[ITEMS]: itemsConfig?.file && [itemsConfig?.file],
			}
			form.setFieldsValue(formData)
		}
	}, [form, projectConfig, predictionsConfig, itemsConfig, annotationsConfig])

	const _onConfigUploadChange = useCallback(
		({ file, fileList }) => {
			if (!isEmpty(fileList)) {
				if (fileList?.length === 2) {
					fileList = fileList.slice(-1)
					form.setFieldsValue({ [CONFIG]: fileList })
					message.warning(t('configurationProject:files.singleFileUpload'))
				}
				const reader = new FileReader()
				reader.readAsText(file)
				reader.onload = () => {
					try {
						const config = JSON.parse(reader.result)
						config.file = fileList

						const overrideConfig =
							!isEmpty(projectConfig) &&
							some(
								config,
								(value, key) =>
									!!projectConfig[key] &&
									!isEqual(value, key === DEADLINE ? projectConfig[key].toISOString() : projectConfig[key])
							)

						if (overrideConfig) {
							Modal.confirm({
								title: t('configurationProject:files.warning'),
								content: t('configurationProject:files.conflict'),
								onOk: () => {
									dispatch(updateConfigProject(FILE, config))
								},
							})
						} else {
							dispatch(updateConfigProject(FILE, config))
						}
					} catch (err) {
						console.error(err)
						message.error(t('configurationProject:files.onlyJsonFile'))
					}
				}
			}
		},
		[t, form, dispatch, projectConfig]
	)

	const _onUploadChange = useCallback(
		(type, isUpdate) =>
			({ file, fileList }) => {
				if (fileList?.length === 2) {
					fileList = fileList.slice(-1)
					form.setFieldsValue({ [type]: fileList })
				}

				switch (type) {
					case ITEMS: {
						dispatch(uploadItems(file, isUpdate))
						break
					}
					case PREDICTIONS: {
						dispatch(uploadPredictions(file))
						break
					}
					case ANNOTATIONS: {
						dispatch(uploadAnnotations(file))
						break
					}
					default:
				}
			},
		[dispatch, form]
	)

	const _onValuesChange = useCallback(
		debounce(
			(value, values) =>
				![CONFIG, PREDICTIONS, ITEMS].includes(Object.keys(value)[0]) && dispatch(updateConfigProject(PROJECT, values)),
			500
		),
		[dispatch]
	)

	const _normFile = (e) => (Array.isArray(e) ? e : e && e.fileList)

	const _beforeUpload = () => false

	return (
		<Styled.Root>
			<Styled.Form layout="vertical" initialValues={projectConfig} form={form} onValuesChange={_onValuesChange}>
				{isEdit && (
					<Styled.AlertContainer>
						<Alert message={t('configurationProject:files.alert')} type="error" showIcon />
					</Styled.AlertContainer>
				)}
				<Styled.ItemsContainer>
					<Styled.Title>{t('configurationProject:files.files')}</Styled.Title>
					<Form.Item
						name={CONFIG}
						label={t('configurationProject:files.configFile')}
						valuePropName="fileList"
						getValueFromEvent={_normFile}
					>
						<Upload name={CONFIG} onChange={_onConfigUploadChange} beforeUpload={_beforeUpload}>
							<Button icon={<UploadOutlined />}>{t('configurationProject:files.add')}</Button>
						</Upload>
					</Form.Item>
					<Form.Item label={t('configurationProject:files.itemFile')} required={!isEdit}>
						<Styled.Space align="start">
							<Form.Item name={ITEMS} valuePropName="fileList" getValueFromEvent={_normFile} noStyle required={!isEdit}>
								<Upload onChange={_onUploadChange(ITEMS, false)} beforeUpload={_beforeUpload} multiple={false}>
									<Button icon={<UploadOutlined />}>{t('configurationProject:files.add')}</Button>
								</Upload>
							</Form.Item>
							{isEdit && (
								<>
									<Styled.Span>{t('configurationProject:files.or')}</Styled.Span>
									<Form.Item
										name={ITEMS}
										valuePropName="fileList"
										getValueFromEvent={_normFile}
										noStyle
										required={!isEdit}
									>
										<Upload
											onChange={_onUploadChange(ITEMS, true)}
											showUploadList={false}
											beforeUpload={_beforeUpload}
											multiple={false}
										>
											<Button icon={<UploadOutlined />}>{t('configurationProject:files.update')}</Button>
										</Upload>
									</Form.Item>
								</>
							)}
						</Styled.Space>
					</Form.Item>
					<Form.Item label={t('configurationProject:files.annotatedFile')}>
						<Space align="start">
							<Form.Item name={ANNOTATIONS} valuePropName="fileList" getValueFromEvent={_normFile} noStyle>
								<Upload onChange={_onUploadChange(ANNOTATIONS)} beforeUpload={_beforeUpload} multiple={false}>
									<Button icon={<UploadOutlined />}>{t('configurationProject:files.add')}</Button>
								</Upload>
							</Form.Item>
						</Space>
					</Form.Item>
					<Form.Item label={t('configurationProject:files.predictionFile')}>
						<Space align="start">
							<Form.Item name={PREDICTIONS} valuePropName="fileList" getValueFromEvent={_normFile} noStyle>
								<Upload onChange={_onUploadChange(PREDICTIONS)} beforeUpload={_beforeUpload} multiple={false}>
									<Button icon={<UploadOutlined />}>{t('configurationProject:files.add')}</Button>
								</Upload>
							</Form.Item>
						</Space>
					</Form.Item>
					{isSimilarityAllowed && (
						<Form.Item label={t('configurationProject:files.api')} name={API}>
							<Styled.Input />
						</Form.Item>
					)}
				</Styled.ItemsContainer>
			</Styled.Form>
		</Styled.Root>
	)
}

export default FilesStepPage
