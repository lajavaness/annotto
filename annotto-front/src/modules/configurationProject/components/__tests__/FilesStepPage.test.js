import { Provider } from 'react-redux'
import { ThemeProvider } from 'styled-components'
import { createStore } from 'redux-dynamic-modules'
import { render } from '@testing-library/react'
import React from 'react'

import theme from '__theme__'

import FilesStepPage from 'modules/configurationProject/components/FilesStepPage'

const initialState = {}
const getStore = (state) => createStore({ initialState: state })
const getInstance = (state = initialState) => (
	<Provider store={getStore(state)}>
		<ThemeProvider theme={theme}>
			<FilesStepPage />
		</ThemeProvider>
	</Provider>
)

describe('FilesStepPage', () => {
	it('matches snapshot', () => {
		const { asFragment } = render(getInstance())
		expect(asFragment()).toMatchSnapshot()
	})
})
