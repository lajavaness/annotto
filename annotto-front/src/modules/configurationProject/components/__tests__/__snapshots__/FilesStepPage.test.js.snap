// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilesStepPage matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-gsBqHK jMCMRZ"
  >
    <form
      class="ant-form ant-form-vertical sc-fybuyZ ddejrc"
    >
      <div
        class="sc-iWFTwQ iupdHB"
      >
        <div
          class="sc-aKXVD hQbhhA"
        >
          configurationProject:files.files
        </div>
        <div
          class="ant-row ant-form-item"
        >
          <div
            class="ant-col ant-form-item-label"
          >
            <label
              class=""
              for="config"
              title="configurationProject:files.configFile"
            >
              configurationProject:files.configFile
            </label>
          </div>
          <div
            class="ant-col ant-form-item-control"
          >
            <div
              class="ant-form-item-control-input"
            >
              <div
                class="ant-form-item-control-input-content"
              >
                <span
                  class=""
                >
                  <div
                    class="ant-upload ant-upload-select ant-upload-select-text"
                  >
                    <span
                      class="ant-upload"
                      role="button"
                      tabindex="0"
                    >
                      <input
                        accept=""
                        id="config"
                        style="display: none;"
                        type="file"
                      />
                      <button
                        class="ant-btn"
                        type="button"
                      >
                        <span
                          aria-label="upload"
                          class="anticon anticon-upload"
                          role="img"
                        >
                          <svg
                            aria-hidden="true"
                            data-icon="upload"
                            fill="currentColor"
                            focusable="false"
                            height="1em"
                            viewBox="64 64 896 896"
                            width="1em"
                          >
                            <path
                              d="M400 317.7h73.9V656c0 4.4 3.6 8 8 8h60c4.4 0 8-3.6 8-8V317.7H624c6.7 0 10.4-7.7 6.3-12.9L518.3 163a8 8 0 00-12.6 0l-112 141.7c-4.1 5.3-.4 13 6.3 13zM878 626h-60c-4.4 0-8 3.6-8 8v154H214V634c0-4.4-3.6-8-8-8h-60c-4.4 0-8 3.6-8 8v198c0 17.7 14.3 32 32 32h684c17.7 0 32-14.3 32-32V634c0-4.4-3.6-8-8-8z"
                            />
                          </svg>
                        </span>
                        <span>
                          configurationProject:files.add
                        </span>
                      </button>
                    </span>
                  </div>
                  <div
                    class="ant-upload-list ant-upload-list-text"
                  />
                </span>
              </div>
            </div>
          </div>
        </div>
        <div
          class="ant-row ant-form-item"
        >
          <div
            class="ant-col ant-form-item-label"
          >
            <label
              class="ant-form-item-required"
              title="configurationProject:files.itemFile"
            >
              configurationProject:files.itemFile
            </label>
          </div>
          <div
            class="ant-col ant-form-item-control"
          >
            <div
              class="ant-form-item-control-input"
            >
              <div
                class="ant-form-item-control-input-content"
              >
                <div
                  class="ant-space ant-space-horizontal ant-space-align-start sc-gIRjGK dNhWuL"
                >
                  <div
                    class="ant-space-item"
                  >
                    <span
                      class=""
                    >
                      <div
                        class="ant-upload ant-upload-select ant-upload-select-text"
                      >
                        <span
                          class="ant-upload"
                          role="button"
                          tabindex="0"
                        >
                          <input
                            accept=""
                            id="items"
                            style="display: none;"
                            type="file"
                          />
                          <button
                            class="ant-btn"
                            type="button"
                          >
                            <span
                              aria-label="upload"
                              class="anticon anticon-upload"
                              role="img"
                            >
                              <svg
                                aria-hidden="true"
                                data-icon="upload"
                                fill="currentColor"
                                focusable="false"
                                height="1em"
                                viewBox="64 64 896 896"
                                width="1em"
                              >
                                <path
                                  d="M400 317.7h73.9V656c0 4.4 3.6 8 8 8h60c4.4 0 8-3.6 8-8V317.7H624c6.7 0 10.4-7.7 6.3-12.9L518.3 163a8 8 0 00-12.6 0l-112 141.7c-4.1 5.3-.4 13 6.3 13zM878 626h-60c-4.4 0-8 3.6-8 8v154H214V634c0-4.4-3.6-8-8-8h-60c-4.4 0-8 3.6-8 8v198c0 17.7 14.3 32 32 32h684c17.7 0 32-14.3 32-32V634c0-4.4-3.6-8-8-8z"
                                />
                              </svg>
                            </span>
                            <span>
                              configurationProject:files.add
                            </span>
                          </button>
                        </span>
                      </div>
                      <div
                        class="ant-upload-list ant-upload-list-text"
                      />
                    </span>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="ant-row ant-form-item"
        >
          <div
            class="ant-col ant-form-item-label"
          >
            <label
              class=""
              title="configurationProject:files.annotatedFile"
            >
              configurationProject:files.annotatedFile
            </label>
          </div>
          <div
            class="ant-col ant-form-item-control"
          >
            <div
              class="ant-form-item-control-input"
            >
              <div
                class="ant-form-item-control-input-content"
              >
                <div
                  class="ant-space ant-space-horizontal ant-space-align-start"
                >
                  <div
                    class="ant-space-item"
                  >
                    <span
                      class=""
                    >
                      <div
                        class="ant-upload ant-upload-select ant-upload-select-text"
                      >
                        <span
                          class="ant-upload"
                          role="button"
                          tabindex="0"
                        >
                          <input
                            accept=""
                            id="annotations"
                            style="display: none;"
                            type="file"
                          />
                          <button
                            class="ant-btn"
                            type="button"
                          >
                            <span
                              aria-label="upload"
                              class="anticon anticon-upload"
                              role="img"
                            >
                              <svg
                                aria-hidden="true"
                                data-icon="upload"
                                fill="currentColor"
                                focusable="false"
                                height="1em"
                                viewBox="64 64 896 896"
                                width="1em"
                              >
                                <path
                                  d="M400 317.7h73.9V656c0 4.4 3.6 8 8 8h60c4.4 0 8-3.6 8-8V317.7H624c6.7 0 10.4-7.7 6.3-12.9L518.3 163a8 8 0 00-12.6 0l-112 141.7c-4.1 5.3-.4 13 6.3 13zM878 626h-60c-4.4 0-8 3.6-8 8v154H214V634c0-4.4-3.6-8-8-8h-60c-4.4 0-8 3.6-8 8v198c0 17.7 14.3 32 32 32h684c17.7 0 32-14.3 32-32V634c0-4.4-3.6-8-8-8z"
                                />
                              </svg>
                            </span>
                            <span>
                              configurationProject:files.add
                            </span>
                          </button>
                        </span>
                      </div>
                      <div
                        class="ant-upload-list ant-upload-list-text"
                      />
                    </span>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="ant-row ant-form-item"
        >
          <div
            class="ant-col ant-form-item-label"
          >
            <label
              class=""
              title="configurationProject:files.predictionFile"
            >
              configurationProject:files.predictionFile
            </label>
          </div>
          <div
            class="ant-col ant-form-item-control"
          >
            <div
              class="ant-form-item-control-input"
            >
              <div
                class="ant-form-item-control-input-content"
              >
                <div
                  class="ant-space ant-space-horizontal ant-space-align-start"
                >
                  <div
                    class="ant-space-item"
                  >
                    <span
                      class=""
                    >
                      <div
                        class="ant-upload ant-upload-select ant-upload-select-text"
                      >
                        <span
                          class="ant-upload"
                          role="button"
                          tabindex="0"
                        >
                          <input
                            accept=""
                            id="predictions"
                            style="display: none;"
                            type="file"
                          />
                          <button
                            class="ant-btn"
                            type="button"
                          >
                            <span
                              aria-label="upload"
                              class="anticon anticon-upload"
                              role="img"
                            >
                              <svg
                                aria-hidden="true"
                                data-icon="upload"
                                fill="currentColor"
                                focusable="false"
                                height="1em"
                                viewBox="64 64 896 896"
                                width="1em"
                              >
                                <path
                                  d="M400 317.7h73.9V656c0 4.4 3.6 8 8 8h60c4.4 0 8-3.6 8-8V317.7H624c6.7 0 10.4-7.7 6.3-12.9L518.3 163a8 8 0 00-12.6 0l-112 141.7c-4.1 5.3-.4 13 6.3 13zM878 626h-60c-4.4 0-8 3.6-8 8v154H214V634c0-4.4-3.6-8-8-8h-60c-4.4 0-8 3.6-8 8v198c0 17.7 14.3 32 32 32h684c17.7 0 32-14.3 32-32V634c0-4.4-3.6-8-8-8z"
                                />
                              </svg>
                            </span>
                            <span>
                              configurationProject:files.add
                            </span>
                          </button>
                        </span>
                      </div>
                      <div
                        class="ant-upload-list ant-upload-list-text"
                      />
                    </span>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
